src/shape_transform_descriptor.cpp

```diff
@@ -0,0 +1,836 @@
+                new_sub.len = start[j];
+                return {{new_sub}};
+            });
+        }
+        r++;
+        i++;
+    }
+
+    // Handle trailing 1s
+    if(new_dims.size() < rdims.size() and not new_dims.empty())
+    {
+        for(auto d : range(rdims.begin() + new_dims.size(), rdims.end()))
+        {
+            if(d != 1)
+                return false;
+            new_dims.push_back({{dimension::sub{1}}});
+        }
+    }
+    assert(rdims.size() == new_dims.size());
+    if(rdims.size() != new_dims.size())
+        return false;
+    dimensions = new_dims;
+    return true;
+}
+
+bool shape_transform_descriptor::apply_transpose(const std::vector<std::int64_t>& permutation)
+{
+    if(permutation.size() != dimensions.size())
+        return false;
+    dimensions = reorder_dims(dimensions, permutation);
+    return true;
+}
+
+bool shape_transform_descriptor::apply_broadcast(const std::vector<std::size_t>& out_lens,
+                                                 optional<std::size_t> axis)
+{
+    auto offset = axis.value_or(out_lens.size() - dimensions.size());
+
+    std::vector<dimension> new_dims;
+    std::transform(out_lens.begin(),
+                   out_lens.begin() + offset,
+                   std::back_inserter(new_dims),
+                   [&](auto len) -> dimension {
+                       return {{dimension::sub{len, {}}}};
+                   });
+    std::transform(dimensions.begin(),
+                   dimensions.end(),
+                   out_lens.begin() + offset,
+                   std::back_inserter(new_dims),
+                   [&](const dimension& dim, auto len) -> dimension {
+                       if(len == dim.len())
+                           return dim;
+                       if(dim.len() != 1)
+                           MIGRAPHX_THROW("Wrong out_lens for broadcast");
+                       return {{dimension::sub{len, {}}}};
+                   });
+    std::transform(out_lens.begin() + offset + dimensions.size(),
+                   out_lens.end(),
+                   std::back_inserter(new_dims),
+                   [&](auto len) -> dimension {
+                       return {{dimension::sub{len, {}}}};
+                   });
+    assert(out_lens.size() == new_dims.size());
+    dimensions = new_dims;
+    return true;
+}
+
+// Remove subdimensions of 1
+void remove_1_sub_dims(std::vector<dimension::sub>& subdimensions)
+{
+    subdimensions.erase(std::remove_if(subdimensions.begin(),
+                                       subdimensions.end(),
+                                       [&](const dimension::sub& d) { return d.len == 1; }),
+                        subdimensions.end());
+}
+
+void dimension::simplify()
+{
+    if(subdimensions.size() < 2)
+        return;
+    remove_1_sub_dims(subdimensions);
+    // Flatten adjacent dimensions
+    adjacent_for_each(subdimensions.begin(), subdimensions.end(), [&](sub& d1, sub& d2) {
+        if(d1.axis.size() < 2)
```

Comment:
For this to work, i think subdimensions need to be in sorted order based on axes.  I do not see it being sorted before calling to `simplify()`

Reply:
They dont need to be sorted. This is checking adjacent axes in a subdimension and making the first one a len of 1(which will be removed). So `{0, 3}, {0, 1}, {0, 2}` becomes `{0, 3}, {0, 2}`(later these will be renumbered to `{0, 2}, {0, 1}`.

Reply:
It is checking if axes in adjacent subdimension differ by 1 at the last position.   For that check to work in all cases, i think it needs to be sorted.

Reply:
> For that check to work in all cases, i think it needs to be sorted.

In what case would it not work?

Reply:
e.g. 
`[[0, 0, 1], [0, 0, 3], [0, 0, 2]]`

Reply:
But how would that not work? Those axes can't be removed.

Reply:
Resolved offline.

subdimensions doesn't need to be sorted. Need to maintain memory layout and therefore only remove ones that are adjacent in memory.